File: 1382-balance-a-binary-search-tree/1382-balance-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    nnums = list()
    visited = list()
    
    def balanced(self, left: int, right: int) -> TreeNode:
        global visited
        if left + 1 == right:
            return TreeNode(val=self.nnums[right], left=TreeNode(val=self.nnums[left]))
        elif left == right:
            return TreeNode(val=self.nnums[right])
        
        mid = (left + right) // 2
        self.visited[mid] = True
        return TreeNode(val=self.nnums[mid], left=self.balanced(left, mid - 1), right=self.balanced(mid + 1, right))
    
    def balanceBST(self, root: TreeNode) -> TreeNode:
        global nnums
        global visited
        queue = [root]
        
        self.nnums = list()
        self.visited = list()
        #노드 Value 뽑기
        node = None
        while len(queue) != 0:
            node = queue[0]
            del queue[0]
            self.nnums.append(node.val)
            self.visited.append(False)
            
            if node.left is not None:
                queue.append(node.left)
            if node.right is not None:
                queue.append(node.right)
                
        self.nnums.sort()
        # print(self.nnums)
        return self.balanced(0, len(self.nnums) - 1)

File: 1382-balance-a-binary-search-tree/1382-balance-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> nnums;
    TreeNode* setNodes(int left, int right) {
        if(left + 1 == right) {
            return new TreeNode(nnums[right], new TreeNode(nnums[left], NULL, NULL), NULL);
        } else if (left == right) {
            return new TreeNode(nnums[right], NULL, NULL);
        }
        int mid = (left + right) / 2;
        return new TreeNode(nnums[mid], setNodes(left, mid - 1), setNodes(mid + 1, right));
    }
    TreeNode* balanceBST(TreeNode* root) {
        vector<TreeNode*> queue;
        queue.push_back(root);
        
        TreeNode* node;
        while(!queue.empty()) {
            node = queue[0];
            queue.erase(queue.begin());
            nnums.push_back(node->val);
            
            if(node->left != NULL) queue.push_back(node->left);
            if(node->right != NULL) queue.push_back(node->right);
        }
        sort(nnums.begin(), nnums.end());
        
        return setNodes(0, nnums.size() - 1);
    }
};